Improved Mathematica parser for multiplication operator

arihantparsoya opened the pull request:
Fixes #8501.

Added support for:
* Space Multiplication of two functions
* `*` operator between a function and a object.

### Sample Program:
```python 
>>> from sympy.parsing.mathematica import mathematica as M
>>> M("x*Sin[1/x]")
x*sin(1/x)
>>> M("Sin[1/x]*x")
x*sin(1/x)
>>> M("Sin[x] Cos[x]")
sin(x)*cos(x)
```

#### Note:

We still need to implement parsing for space multiplication. Some [work](https://github.com/sympy/sympy/pull/8766/files#diff-5bead3fe0fbe39640e6bbc0055461aefR45) is done by @nishithshah2211 on space multiplication but that solution will not always be true.

I think we need better framework for parsing than `Mathematica` parser already has (such as parsing using ASTs) to implement space multiplication.

arihantparsoya committed:
Fixes #8501.

Added support for:

1. Space Multiplication of two functions
2. `*` operator between a function and a object.

parsoyaarihant commented:
@certik , can you review this?

Upabjojr commented on sympy/parsing/mathematica.py:
```diff
@@ -23,6 +23,10 @@ def parse(s):
             m.group(2)) + parse(m.group(3))),
 
+        # Space multiplication between two functions
+        (r"\A(\w+\[[^\]]+[^\[]*\])( )(\w+\[[^\]]+[^\[]*\])\Z",
```
plus after the space?

parsoyaarihant replied:
This pattern only matches space between two functions:
```python
>>> from re import match
>>> p = r"\A(\w+\[[^\]]+[^\[]*\])( )(\w+\[[^\]]+[^\[]*\])\Z"
>>> s = 'Sin[x] +Cos[x]'
>>> print(match(p,s))
None
```

Upabjojr replied:
I mean, many spaces, like: `Sin[x]    Cos[x]`. I think it's still a valid Mathematica expression.

parsoyaarihant replied:
Fixed it.

Upabjojr commented on sympy/parsing/mathematica.py:
```diff
@@ -42,6 +46,10 @@ def parse(s):
         lambda m: parse(m.group(1)) + "*" + parse(m.group(2))),
 
+        # Multiplication between any object and a function
+        (r"\A(.+)([*])(\w+\[[^\]]+[^\[]*\])\Z",
```
Should space also be supported here?

parsoyaarihant replied:
Suppose I use pattern
```python 
r"\A(.+)([ ])(\w+\[[^\]]+[^\[]*\])\Z"
```
Then it will also match functions such as `x + Sin[x]` (into `x + ` and `Sin[x]`). I think we need an algorithm to detect when `space` imply `multiplication`.

Upabjojr replied:
What about `a Sin[x]`? If I recall correctly, there is some way of putting alternative matching conditions in the regex.

parsoyaarihant replied:
Implemented it.

I [implemented](https://github.com/sympy/sympy/pull/12528/files#diff-5bead3fe0fbe39640e6bbc0055461aefR17) arithmetic operator for containing spaces( such as `a + b`) before implementing space multiplication so these expressions get matched before the parser matches for space multiplication.

arihantparsoya committed:
implemented space multiplication

-> implemented space multiplication between any object and a
function(line 54 of mathematica.py).

-> added support for whitespace in space multiplication between two
functions in line 27.

-> improved arithmetic operation between a constant and a function to
account for whitespaces.

Upabjojr commented:
I've just noticed this:
https://github.com/rljacobson/FoxySheep

parsoyaarihant commented:
Thanks for sharing this. I am exploring FoxySheep and will try to write a parser using it.

Upabjojr commented:
@parsoyaarihant I have to idea what the quality of FoxySheep is. The license is compatible with SymPy's (BSD 2-clause).

czgdp1807 commented:
Closing as per https://github.com/sympy/sympy/issues/8501#issuecomment-558121546 and https://github.com/sympy/sympy/issues/8501#issuecomment-534188286
Thanks for the contributions. Please let me know if it is required to be opened again.
